search/src/java/cz/incad/Kramerius/oai/OaiServlet.java

```diff
@@ -109,16 +109,31 @@ protected void doGet(HttpServletRequest req, HttpServletResponse resp) throws Se
         try {
             sendResponse(oaiWriter, resp);
         } catch (IOException ex) {
-            LOG.log(Level.SEVERE, pid, ex);
+            LOG.log(Level.WARNING, pid + " doesn't exist in Fedora.");
             if (!resp.isCommitted()) {
                 resp.reset();
-                resp.sendError(HttpServletResponse.SC_INTERNAL_SERVER_ERROR);
+                sendFakeContent(resp);
             }
         } finally {
             oaiWriter.close();
         }
 
     }
+    
+    private void sendFakeContent(HttpServletResponse resp) {
+        String emptyXml = "<error code=\"404\">";
+        resp.setContentType("text/xml");
+        resp.setCharacterEncoding("UTF-8");
+        resp.setContentLength(emptyXml.length());
+
+        try {
+            resp.getWriter().write(emptyXml);
+            resp.getWriter().flush();
+            resp.getWriter().close();
```

Comment:
Close - zavírat zdroje bys měl vždy v bloku finally. Kramerius už konečně opustil JDK 6, takže můžeme používat novější konstrukty. Např. try-with-resources, PrintWriter implementuje rozhraní closeable, takže by to mělo jít.